File: apps/expo/modules/streamer/ios/ThumbnailGenerator.swift
import Foundation
import ZIPFoundation
import UIKit
import os.log
import Readium

/// Generates thumbnails from comic book archives and ebooks
class ThumbnailGenerator {
    static let shared = ThumbnailGenerator()
    
    private let logger = Logger(subsystem: "com.stump.streamer", category: "thumbnail")
    
    private let defaultThumbnailSize: CGFloat = 300
    
    private init() {}
    
    /// Generate a thumbnail from the first valid page of an archive or ebook cover
    /// - Parameters:
    ///   - bookId: The book ID to use for the thumbnail filename
    ///   - archivePath: Path to the archive file
    ///   - outputDir: Directory where {bookId}.jpg will be saved
    /// - Throws: ThumbnailError if generation fails
    func generateThumbnail(
        bookId: String,
        archivePath: String,
        outputDir: String
    ) throws {
        logger.info("Generating thumbnail for book: \(bookId)")
        
        try createOutputDirectoryIfNeeded(at: outputDir)
        
        let outputPath = (outputDir as NSString).appendingPathComponent("\(bookId).jpg")
        
        let fileExtension = (archivePath as NSString).pathExtension.lowercased()
        let isReadiumRequired = fileExtension == "epub" || fileExtension == "pdf"
        
        let imageData: Data
        
        if isReadiumRequired {
            imageData = try extractReadiumCover(archivePath: archivePath)
        } else {
            imageData = try extractZipCover(archivePath: archivePath)
        }
        
        guard let originalImage = UIImage(data: imageData) else {
            throw ThumbnailError.invalidImageData
        }
        
        let scaledImage = try scaleImage(originalImage, maxSize: defaultThumbnailSize)
        
        guard let thumbnailData = scaledImage.jpegData(compressionQuality: 0.8) else {
            throw ThumbnailError.imageEncodingFailed
        }
        
        let outputURL = URL(fileURLWithPath: outputPath)
        try thumbnailData.write(to: outputURL)
        
        logger.info("Thumbnail generated: \(outputPath)")
    }
    
    /// Extract cover from a standard archive (CBZ, ZIP)
    private func extractZipCover(archivePath: String) throws -> Data {
        let archive = try ZipArchive(path: archivePath)
        let imageFiles = archive.getImageFiles()
        
        guard !imageFiles.isEmpty else {
            throw ThumbnailError.noValidImages
        }
        
        let firstImageEntry = imageFiles[0]
        logger.debug("Using first image from comic: \(firstImageEntry.path)")
        
        return try archive.extractEntry(firstImageEntry)
    }
    
    /// Extract cover from an EPUB/PDF, falling back to ZIP extraction if needed
    private func extractReadiumCover(archivePath: String) throws -> Data {
        let fileURL = URL(fileURLWithPath: archivePath)
        
        // TODO: Determine if better to make thumb gen async
        // Note: I did this to keep things synchronous
        var extractedData: Data?
        let semaphore = DispatchSemaphore(value: 0)
        
        Task {
            extractedData = await BookService.instance.getCoverImage(from: fileURL)
            semaphore.signal()
        }
        
        semaphore.wait()
        
        if let data = extractedData {
            logger.debug("Successfully extracted cover: \(data.count) bytes")
            return data
        }
        
        logger.warning("Failed to extract cover, falling back to ZIP extraction")
        if archivePath.lowercased().hasSuffix(".pdf") {
            throw ThumbnailError.pdfCoverExtractionFailed
        }
        return try extractZipCover(archivePath: archivePath)
    }
    
    private func scaleImage(_ image: UIImage, maxSize: CGFloat) throws -> UIImage {
        let originalSize = image.size
        
        let scaleFactor = min(maxSize / originalSize.width, maxSize / originalSize.height)
        
        // Don't need to scale if already within size
        if scaleFactor >= 1.0 {
            return image
        }
        
        let newSize = CGSize(
            width: originalSize.width * scaleFactor,
            height: originalSize.height * scaleFactor
        )
        
        let renderer = UIGraphicsImageRenderer(size: newSize)
        let scaledImage = renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: newSize))
        }
        
        return scaledImage
    }
    
    private func createOutputDirectoryIfNeeded(at path: String) throws {
        let url = URL(fileURLWithPath: path)
        if !FileManager.default.fileExists(atPath: path) {
            do {
                try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
            } catch {
                throw ThumbnailError.outputDirectoryCreationFailed(error)
            }
        }
    }
}

enum ThumbnailError: Error {
    case noValidImages
    case invalidImageData
    case imageEncodingFailed
    case outputDirectoryCreationFailed(Error)
    case epubCoverExtractionFailed(String)
    case pdfCoverExtractionFailed
    
    var localizedDescription: String {
        switch self {
        case .noValidImages:
            return "No valid images found in archive"
        case .invalidImageData:
            return "Could not create image from data"
        case .imageEncodingFailed:
            return "Failed to encode image as JPEG"
        case .outputDirectoryCreationFailed(let error):
            return "Failed to create output directory: \(error.localizedDescription)"
        case .epubCoverExtractionFailed(let reason):
            return "Failed to extract EPUB cover: \(reason)"
        case .pdfCoverExtractionFailed:
            return "Failed to extract PDF cover"
        }
    }
}

File: apps/expo/modules/streamer/android/src/main/java/expo/modules/streamer/ThumbnailGenerator.kt
package expo.modules.streamer

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import expo.modules.readium.BookService
import kotlinx.coroutines.runBlocking
import java.io.File
import java.io.FileOutputStream
import java.net.URL

/**
 * Errors that can occur during thumbnail generation
 */
sealed class ThumbnailError : Exception() {
    object NoValidImages : ThumbnailError() {
        override val message: String get() = "No valid image files found in archive"
    }

    data class ExtractionFailed(override val cause: Throwable) : ThumbnailError() {
        override val message: String get() = "Failed to extract image from archive: ${cause.message}"
    }

    data class DecodingFailed(override val cause: Throwable) : ThumbnailError() {
        override val message: String get() = "Failed to decode image to bitmap: ${cause.message}"
    }

    data class SavingFailed(override val cause: Throwable) : ThumbnailError() {
        override val message: String get() = "Failed to save thumbnail: ${cause.message}"
    }
}

/**
 * Generates thumbnails from comic book archives and ebooks
 */
class ThumbnailGenerator private constructor(private val context: Context) {

    companion object {
        private const val TAG = "ThumbnailGenerator"
        // Note: This differs from iOS implementation since Swift automagically handles scaling for us
        private const val MAX_THUMBNAIL_SIZE_DP = 300
        private const val JPEG_QUALITY = 80

        // Note: This took me forever to figure out, I am happy to not be a Kotlin dev lol
        @Volatile
        private var INSTANCE: ThumbnailGenerator? = null

        fun getInstance(context: Context): ThumbnailGenerator {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: ThumbnailGenerator(context.applicationContext).also {
                    INSTANCE = it
                }
            }
        }

        private fun stripFilePrefix(path: String): String {
            return if (path.startsWith("file://")) {
                path.substring(7)
            } else {
                path
            }
        }
    }

    private val bookService by lazy { BookService(context) }
    
    /**
     * Convert dp to pixels based on the device's density
     */
    private val maxThumbnailSizePixels: Int by lazy {
        val density = context.resources.displayMetrics.density
        (MAX_THUMBNAIL_SIZE_DP * density).toInt()
    }

    /**
     * Generate a thumbnail for a book
     * 
     * @param bookId Unique identifier for the book
     * @param archivePath Path to the archive file
     * @param outputDir Directory to save the thumbnail
     * @return Path to the generated thumbnail
     * @throws ThumbnailError if generation fails
     */
    fun generateThumbnail(
        bookId: String,
        archivePath: String,
        outputDir: String
    ): String {
        val cleanArchivePath = stripFilePrefix(archivePath)
        val cleanOutputDir = stripFilePrefix(outputDir)

        Log.d(TAG, "Generating thumbnail for book $bookId")

        val outputDirectory = File(cleanOutputDir)
        if (!outputDirectory.exists()) {
            outputDirectory.mkdirs()
        }

        val fileExtension = File(cleanArchivePath).extension.lowercase()
        val isReadiumRequired = fileExtension == "epub" || fileExtension == "pdf"

        val bitmap = if (isReadiumRequired) {
            extractReadiumCover(cleanArchivePath)
        } else {
            extractZipCover(cleanArchivePath)
        }

        val scaledBitmap = scaleBitmap(bitmap)
        
        if (bitmap != scaledBitmap) {
            bitmap.recycle()
        }

        val thumbnailPath = getThumbnailPath(bookId, cleanOutputDir)
        try {
            FileOutputStream(thumbnailPath).use { stream ->
                scaledBitmap.compress(Bitmap.CompressFormat.JPEG, JPEG_QUALITY, stream)
            }
            Log.d(TAG, "Thumbnail saved to: $thumbnailPath")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save thumbnail", e)
            throw ThumbnailError.SavingFailed(e)
        } finally {
            scaledBitmap.recycle()
        }

        return thumbnailPath
    }

    /**
     * Extract cover from a zip archive (CBZ, ZIP)
     */
    private fun extractZipCover(archivePath: String): Bitmap {
        val archive = try {
            ZipArchive(archivePath)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to open archive", e)
            throw ThumbnailError.ExtractionFailed(e)
        }

        val imageFiles = archive.getImageFiles()
        if (imageFiles.isEmpty()) {
            archive.close()
            throw ThumbnailError.NoValidImages
        }

        val firstImage = imageFiles[0]
        Log.d(TAG, "Using first image: ${firstImage.name}")

        val imageData = try {
            archive.extractEntry(firstImage)
        } catch (e: Exception) {
            archive.close()
            Log.e(TAG, "Failed to extract image", e)
            throw ThumbnailError.ExtractionFailed(e)
        } finally {
            archive.close()
        }

        return try {
            BitmapFactory.decodeByteArray(imageData, 0, imageData.size)
                ?: throw IllegalStateException("BitmapFactory returned null")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to decode image", e)
            throw ThumbnailError.DecodingFailed(e)
        }
    }

    /**
     * Extract cover from an EPUB/PDF, falling back to ZIP extraction if needed
     */
    private fun extractReadiumCover(archivePath: String): Bitmap {
        return runBlocking {
            val fileUrl = URL("file://$archivePath")
            
            val coverBitmap = bookService.getCoverImage(fileUrl)
            
            if (coverBitmap != null) {
                Log.d(TAG, "Successfully extracted cover: ${coverBitmap.width}x${coverBitmap.height}")
                coverBitmap
            } else {
                if (archivePath.lowercase().endsWith(".pdf")) {
                    throw ThumbnailError.ExtractionFailed(
                        IllegalStateException("PDF cover extraction failed")
                    )
                }
                Log.w(TAG, "Failed to extract cover, falling back to ZIP extraction")
                extractZipCover(archivePath)
            }
        }
    }

    fun getThumbnailPath(bookId: String, outputDir: String): String {
        val cleanOutputDir = stripFilePrefix(outputDir)
        return File(cleanOutputDir, "$bookId.jpg").absolutePath
    }

    /**
     * Scale a bitmap to fit within maxThumbnailSizePixels while maintaining aspect ratio
     */
    private fun scaleBitmap(original: Bitmap): Bitmap {
        val width = original.width
        val height = original.height

        if (width <= maxThumbnailSizePixels && height <= maxThumbnailSizePixels) {
            return original
        }

        val scale = minOf(
            maxThumbnailSizePixels.toFloat() / width.toFloat(),
            maxThumbnailSizePixels.toFloat() / height.toFloat()
        )

        val newWidth = (width * scale).toInt()
        val newHeight = (height * scale).toInt()

        Log.d(TAG, "Scaling bitmap from ${width}x${height} to ${newWidth}x${newHeight}")

        val scaledBitmap = Bitmap.createBitmap(newWidth, newHeight, Bitmap.Config.ARGB_8888)
        val canvas = android.graphics.Canvas(scaledBitmap)
        
        val paint = android.graphics.Paint().apply {
            isAntiAlias = true
            isFilterBitmap = true
            isDither = true
        }
        
        val srcRect = android.graphics.Rect(0, 0, width, height)
        val dstRect = android.graphics.Rect(0, 0, newWidth, newHeight)
        
        canvas.drawBitmap(original, srcRect, dstRect, paint)
        
        return scaledBitmap
    }
}
